modules/post/linux/manage/adduser.rb

```diff
@@ -0,0 +1,274 @@
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+
+require 'unix_crypt'
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::File
+  include Msf::Post::Unix
+  include Msf::Post::Linux::System
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Add a new user to the system',
+        'Description' => %q{
+          This command adds a new user to the system
+        },
+        'License' => MSF_LICENSE,
+        'Author' => ['Nick Cottrell <ncottrellweb[at]gmail.com>'],
+        'Platform' => ['linux', 'unix', 'bsd', 'aix', 'solaris'],
+        'Privileged' => false,
+        'SessionTypes' => %w[meterpreter shell],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [],
+          'SideEffects' => [CONFIG_CHANGES]
+        }
+      )
+    )
+    register_options([
+      OptString.new('USERNAME', [ true, 'The username to create', 'metasploit' ]),
+      OptString.new('PASSWORD', [ true, 'The password for this user', 'Metasploit$1' ]),
+      OptString.new('SHELL', [true, 'Set the shell that the new user will use', '/bin/sh']),
+      OptString.new('HOME', [true, 'Set the home directory of the new user. Leave empty if user will have no home directory', '']),
+      OptString.new('GROUPS', [false, 'Set what groups the new user will be part of separated with a space', ''])
+    ])
+
+    register_advanced_options([
+      OptEnum.new('UseraddMethod', [true, 'Set how the module adds in new users and groups. AUTO will autodetect how to add new users, MANUAL will add users without any binaries, and CUSTOM will attempt to use a custom designated binary', 'AUTO', ['AUTO', 'MANUAL', 'CUSTOM']]),
+      OptString.new('UseraddBinary', [false, 'Set binary used to set password if you dont want module to find it for you. Set this to \'MANUAL\' to run without binary', nil]),
+      OptEnum.new('SudoMethod', [false, 'Set the method that the new user can obtain root. SUDO_FILE adds the user directly to sudoers while GROUP adds the new user to the sudo group', 'GROUP', ['SUDO_FILE', 'GROUP', 'NONE']]),
+      OptEnum.new('MissingGroups', [true, 'Set how nonexisting groups are handled on the system. Either give an error in the module, ignore it and throw it out, or create the group on the system.', 'ERROR', ['ERROR', 'IGNORE', 'CREATE']]),
+    ])
+  end
+
+  def check_group_exists?(group_name, group_data)
+    return group_data =~ /^#{group_name}:/
+  end
+
+  def d_cmd_exec(command)
+    vprint_status(command)
+    print_line(cmd_exec(command))
+  end
+
+  def run
+    fail_with(Failure::NoAccess, 'Session isnt running as root') unless is_root?
+    case datastore['UseraddMethod']
+    when 'CUSTOM'
+      fail_with(Failure::NotFound, "Cannot find command on path given: #{datastore['UseraddBinary']}") unless command_exists?(datastore['UseraddBinary'])
```

Comment:
Huh. thats a great point. I may have gotten a good amount of experience with ruby, but how do you think I should handle that exception? After all, I'm giving a fail if the `command_exists?` command fails anyway.

Reply:
Since `command_exists?` is called multiple times in the module, it would make more sense to create a separate wrapper method (let's say `check_command_exists?(cmd)` for example) that would call the proper `command_exists?` protected by a `rescue` block. For example:
```ruby
def check_command_exists?(cmd)
  command_exists?(cmd)
rescue RuntimeError => e
  fail_with(Failure::Unknown, "Unable to check if the command `#{cmd}` exists: #{e}")
end
```

I believe if this command fails, this means there is an issue with the actual session (disconnection, permissions, etc.) and failing the entire module looks like a reasonable option, but it is up to you if you want to handle this differently.

Also, we need to `rescue` the generic `RuntimeError` since, unfortunately, the method is [raising](https://github.com/rapid7/metasploit-framework/blob/40a3ce6cdea58a711368369cced70e119520b001/lib/msf/core/post/common.rb#L247) without specifying the exception class.

Reply:
Cool beans. It's been implemented here 85fc0f48e78465d85893db2cece85f097c4000c6